### PR TITLE
Pr l1tx unpackers

### PR DESCRIPTION
**Pull request by mulhearn**
Contains the unpackers for the CaloLayer1 and TwinMux subdetectors.

### CODE REVIEW

**Comment by mulhearn** [comment]
please test

**Comment by mulhearn** [comment]
+1

**Comment by mulhearn** [comment]
@davidlange6  this one is harmless (not used anywhere yet) and compiles.  Can you merge this one?

File: EventFilter/L1TXRawToDigi/plugins/L1TCaloLayer1RawToDigi.cc [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+}
+
+void L1TCaloLayer1RawToDigi::makeECalTPGs(uint32_t lPhi, UCTCTP7RawData& ctp7Data, std::auto_ptr<EcalTrigPrimDigiCollection>& ecalTPGs) {
+  UCTCTP7RawData::CaloType cType = UCTCTP7RawData::EBEE;
+  for(uint32_t iPhi = 0; iPhi < 4; iPhi++) { // Loop over all four phi divisions on card
+    int cPhi = - 1 + lPhi * 4 + iPhi; // Calorimeter phi index
+    if(cPhi == 0) cPhi = 72;
+    else if(cPhi == -1) cPhi = 71;
+    else if(cPhi < -1) {
+      std::cerr << "L1TCaloLayer1RawToDigi: Major error in makeECalTPGs" << std::endl;
```
Comment (davidlange6) [inline EventFilter/L1TXRawToDigi/plugins/L1TCaloLayer1RawToDigi.cc]: please use logerror..

File: EventFilter/L1TXRawToDigi/plugins/L1TCaloLayer1RawToDigi.cc [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+}
+
+void L1TCaloLayer1RawToDigi::makeHCalTPGs(uint32_t lPhi, UCTCTP7RawData& ctp7Data, std::auto_ptr<HcalTrigPrimDigiCollection>& hcalTPGs) {
+  UCTCTP7RawData::CaloType cType = UCTCTP7RawData::HBHE;
+  for(uint32_t iPhi = 0; iPhi < 4; iPhi++) { // Loop over all four phi divisions on card
+    int cPhi = - 1 + lPhi * 4 + iPhi; // Calorimeter phi index
+    if(cPhi == 0) cPhi = 72;
+    else if(cPhi == -1) cPhi = 71;
+    else if(cPhi < -1) {
+      std::cerr << "L1TCaloLayer1RawToDigi: Major error in makeHCalTPGs" << std::endl;
```
Comment (davidlange6) [inline EventFilter/L1TXRawToDigi/plugins/L1TCaloLayer1RawToDigi.cc]: and here

File: EventFilter/L1TXRawToDigi/plugins/L1TTwinMuxRawToDigi.cc [hunk 3]
```diff
@@ -0,0 +1,693 @@
...
+    throw cms::Exception("TwinMux_unpacker") << "Configuration file error. Size of \'wheels\' and \'feds\' differs.\n";  
+
+  if ( amcsecmap_.size() != wheels_.size() )
+    throw cms::Exception("TwinMux_unpacker") << "Configuration file error. Size of \'wheels\' and \'amcsecmap\' differs.\n";  
+    
+  for (size_t wh_i = 0; wh_i < amcsecmap_.size(); ++wh_i){
+    std::array<short, 12> whmap;      
+    for (size_t amc_i = 1; amc_i < 13; ++amc_i ){
+      short shift = (12-amc_i)*4;
+      whmap.at(amc_i-1) = ( amcsecmap_.at(wh_i) >> shift ) & 0xF;
```
Comment (davidlange6) [inline EventFilter/L1TXRawToDigi/plugins/L1TTwinMuxRawToDigi.cc]: try to use [] not at

File: EventFilter/L1TXRawToDigi/plugins/UCTCTP7RawData.h [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+
+  uint32_t getIndex(CaloType cType, bool negativeEta, uint32_t cEta, uint32_t iPhi) {
+    uint32_t index = 0xDEADBEEF;
+    if(cType == EBEE || cType == HBHE) {
+      if(iPhi > 3) {
+	std::cerr << "Incorrect iPhi; iPhi = " << iPhi << "; should be in [0,3]" << std::endl;
+	return 0xDEADBEEF;
+      }
+      if(cEta < 1 || cEta > 28) {
+	std::cerr << "Incorrect caloEta; cEta = " << cEta << "; should be in [1-28]" << std::endl;
```
Comment (davidlange6) [inline EventFilter/L1TXRawToDigi/plugins/UCTCTP7RawData.h]: logerror..

File: EventFilter/L1TXRawToDigi/python/caloLayer1Stage2Digis_cfi.py [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import FWCore.ParameterSet.Config as cms
+
+l1tCaloLayer1Digis = cms.EDProducer(
+    'L1TCaloLayer1RawToDigi',
+    fedRawDataLabel = cms.InputTag("rawDataCollector"),
+    FEDIDs = cms.vint32(1354, 1356, 1358),
```
Comment (davidlange6) [inline EventFilter/L1TXRawToDigi/python/caloLayer1Stage2Digis_cfi.py]: in longer term, the fed-id lists usually come from the data formats headers for FEDRawData

**Comment by mulhearn** [comment]
@jpazzini if you would like to make these fixes, be my guest.  You are a cms-l1t-offline member so you can push directly back to this branch.  Use CMSSW_8_0_0_pre5 so you won't have to bother with rebasing...  Let me know if you have questions.  Otherwise, I'll get to this eventually.

**Comment by davidlange6** [comment]
@cmsbuild, please test

**Comment by mulhearn** [comment]
Jenkins, have you been drinking?

**Comment by mulhearn** [comment]
please test

**Comment by mulhearn** [comment]
please test

**Comment by mulhearn** [comment]
+1

**Comment by mulhearn** [comment]
@davidlange6 please don't forget this guy, which is easily missed since no l1 sign off needed (new package).

**Comment by jpazzini** [comment]
@mulhearn @davidlange6 - spotted a bug on the TwinMux unpacker after the MWGR. I have the bug-fix ready to be pushed. How do I behave? Do we need to make a new PR, or should we re-open this one?

**Comment by perrotta** [comment]
@jpazzini: this PR has been merged already, please make a new one

**Comment by deguio** [comment]
ciao,
not sure how to read you otherwise.
I see that in this PR, wrt #13058, the 
EventFilter/TwinMuxRawToDigi/python/dttmunpacker_cfi.py
is missing.
however this is requested by the online DT application that went in with #13123 which is now not able to run.

what is the plan?
thanks,
F.

**Comment by jpazzini** [comment]
@deguio - with respect to the "original" PR, the TwinMux unpacker changed its name when it was requested to merge it into this branch. I thought this decision was taken together by DT and L1. @davidlange6 might know more about this.

**Pull request merged**